### PR TITLE
EN-144: Allow to Cancel leave request to keep track

### PR DESCRIPTION
**Pull request by FedericoA15**
# Description :pencil:

A post type endpoint was created to change the status of the pt, status types were added, and when the pt is canceled it re-credits the vacation for the specific employee

## Link to ticket :link:

https://entropyteam.atlassian.net/browse/EN-144

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested? :microscope:

Test 1°
image

Test 2°
image

They are tests of services in vacations

### CODE REVIEW

File: src/main/java/com/entropyteam/entropay/employees/services/PtoService.java [hunk 1]
```diff
@@ -192,4 +193,20 @@ private CalendarEventDto createCalendarEventDto(Pto pto){
         String eventName = employee.getFirstName() + " " + employee.getLastName() + " " + leaveType.getName();
         return new CalendarEventDto(eventId, eventName, startDate, endDate);
     }
+
+    public PtoDto cancelPto(UUID id, PtoDto ptoDto) {
+        Pto oldEntity = ptoRepository.findById(id).orElseThrow();
+        Pto entityToUpdate = toEntity(ptoDto);
+        entityToUpdate.setId(id);
+        entityToUpdate.setStatus(entityToUpdate.getStatus());
```
Comment (agustinb4) [inline src/main/java/com/entropyteam/entropay/employees/services/PtoService.java:201]: for the cancellation you just need to set status CANCELLED to the oldEntity (probably a better name):
`entityToUpdate.setStatus(Status.CANCELLED);`

File: src/main/java/com/entropyteam/entropay/employees/services/PtoService.java [hunk 2]
```diff
@@ -192,4 +193,20 @@ private CalendarEventDto createCalendarEventDto(Pto pto){
...
+        if (isVacationType(oldEntity) && isVacationType(entityToUpdate) && oldEntity.getStatus().equals(Status.APPROVED)
```
Comment (agustinb4) [inline src/main/java/com/entropyteam/entropay/employees/services/PtoService.java:202]: here we only need to ask for `isVacationType(oldEntity)`

File: src/main/java/com/entropyteam/entropay/employees/services/PtoService.java [hunk 3]
```diff
@@ -192,4 +193,20 @@ private CalendarEventDto createCalendarEventDto(Pto pto){
...
+                && ptoDto.status().equals(Status.CANCELLED)) {
+            vacationService.addVacationCredit(entityToUpdate.getEmployee(), entityToUpdate.getDaysAsInteger());
+            googleService.deleteGoogleCalendarEvent(id.toString());
```
Comment (agustinb4) [inline src/main/java/com/entropyteam/entropay/employees/services/PtoService.java:205]: in all cases we need to delete the google calendar event

File: src/main/java/com/entropyteam/entropay/employees/services/PtoService.java [hunk 4]
```diff
@@ -192,4 +193,20 @@ private CalendarEventDto createCalendarEventDto(Pto pto){
...
+        }
+        Pto savedEntity = getRepository().save(entityToUpdate);
+        CalendarEventDto calendarEventDto = createCalendarEventDto(savedEntity);
+        googleService.updateGoogleCalendarEvent(calendarEventDto);
```
Comment (agustinb4) [inline src/main/java/com/entropyteam/entropay/employees/services/PtoService.java:209]: remove those lines, we don't have to update events, only delete

File: src/main/java/com/entropyteam/entropay/employees/services/VacationService.java [hunk 5]
```diff
@@ -107,4 +107,30 @@ protected Vacation toEntity(VacationDto entity) {
 
         return vacation;
     }
+    public void addVacationCredit(Employee employee, Integer totalDaysToAdd) {
```
Comment (agustinb4) [inline src/main/java/com/entropyteam/entropay/employees/services/VacationService.java:110]: please remove this method and use discountVacationDebit to restore employee vacation days

**Review by agustinb4 — CHANGES_REQUESTED** [review]

File: src/main/java/com/entropyteam/entropay/employees/controllers/PtoController.java [hunk 6]
```diff
@@ -5,22 +5,37 @@
...
+import org.springframework.web.bind.annotation.CrossOrigin;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.RequestBody;
 
 @RestController
 @CrossOrigin
 @Secured({ROLE_MANAGER_HR, ROLE_ADMIN, ROLE_DEVELOPMENT})
```
[the remark this answers is not in the document]
Comment (FedericoA15) [reply src/main/java/com/entropyteam/entropay/employees/controllers/PtoController.java]: Resolve

File: src/main/java/com/entropyteam/entropay/employees/controllers/PtoController.java [hunk 7]
```diff
@@ -5,22 +5,37 @@
...
-    public PtoController(CrudService<PtoDto, UUID> crudService) {
+    private final PtoService ptoService;
+
+    public PtoController(CrudService<PtoDto, UUID> crudService, PtoService ptoService) {
         super(crudService);
+        this.ptoService = ptoService;
+    }
+
+    @PostMapping("/{id}/cancel")
+    @Secured({ROLE_MANAGER_HR, ROLE_ADMIN, ROLE_DEVELOPMENT})
```
[the remark this answers is not in the document]
Comment (FedericoA15) [reply src/main/java/com/entropyteam/entropay/employees/controllers/PtoController.java]: Resolve

File: src/main/java/com/entropyteam/entropay/employees/controllers/PtoController.java [hunk 8]
```diff
@@ -5,22 +5,37 @@
...
+    public ResponseEntity<PtoDto> cancelStatus(@PathVariable UUID id, @RequestBody PtoDto ptoDto) {
```
[the remark this answers is not in the document]
Comment (FedericoA15) [reply src/main/java/com/entropyteam/entropay/employees/controllers/PtoController.java]: Resolve

File: src/main/java/com/entropyteam/entropay/employees/models/Pto.java [hunk 9]
```diff
@@ -35,11 +36,11 @@ public class Pto extends BaseEntity {
     @Column
     private Integer labourHours;
 
-    @ManyToOne(fetch = FetchType.LAZY)
+    @ManyToOne(fetch = FetchType.EAGER)
```
[the remark this answers is not in the document]
Comment (FedericoA15) [reply src/main/java/com/entropyteam/entropay/employees/models/Pto.java]: I fix it there using @transactional

File: src/main/java/com/entropyteam/entropay/employees/services/PtoService.java (hunk 1 above)
Comment (FedericoA15) [reply src/main/java/com/entropyteam/entropay/employees/services/PtoService.java]: Resolve

File: src/main/java/com/entropyteam/entropay/employees/services/PtoService.java [hunk 10]
```diff
@@ -90,7 +89,7 @@ public PtoDto update(UUID id, PtoDto ptoDto) {
             vacationService.addVacationDebit(entityToUpdate.getEmployee(), entityToUpdate.getDaysAsInteger());
         }
         entityToUpdate.setId(id);
-        entityToUpdate.setStatus(Status.APPROVED); // For now all approved
+        entityToUpdate.setStatus(entityToUpdate.getStatus());
```
[the remark this answers is not in the document]
Comment (FedericoA15) [reply src/main/java/com/entropyteam/entropay/employees/services/PtoService.java]: Resolved

File: src/main/java/com/entropyteam/entropay/employees/controllers/PtoController.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (FedericoA15) [reply src/main/java/com/entropyteam/entropay/employees/controllers/PtoController.java]: Resolved

File: src/main/java/com/entropyteam/entropay/employees/services/PtoService.java (hunk 2 above)
Comment (FedericoA15) [reply src/main/java/com/entropyteam/entropay/employees/services/PtoService.java]: Resolved

File: src/main/java/com/entropyteam/entropay/employees/services/PtoService.java (hunk 3 above)
Comment (FedericoA15) [reply src/main/java/com/entropyteam/entropay/employees/services/PtoService.java]: Resolved

File: src/main/java/com/entropyteam/entropay/employees/services/VacationService.java (hunk 5 above)
Comment (FedericoA15) [reply src/main/java/com/entropyteam/entropay/employees/services/VacationService.java]: Resolved

File: src/main/java/com/entropyteam/entropay/employees/services/PtoService.java (hunk 4 above)
Comment (FedericoA15) [reply src/main/java/com/entropyteam/entropay/employees/services/PtoService.java]: Resolved

**Review by agustinb4 — APPROVED** [review]
Looks good!

**Review by jauday — APPROVED** [review]

**Pull request merged**